### PR TITLE
Mathesar 946 record preview data

### LINKED ISSUE #946: Modify records API to return preview data for foreign key columns
## Problem

- For cells in columns with foreign keys, the front end will sometimes need to display a preview of the referenced record instead of the value stored in the column.
- Separate API calls to retrieve that record preview data would result in poor performance because we'd need to issue one call per row. Ideally the records API would return that data within the same response as the table data.

## Solution
- The records API should accept a query parameter (perhaps something like `fk_previews`) to allow clients to toggle whether to return the preview data for foreign keys, (along with the underlying column value too). 
    - `fk_previews` should support three values:
        - `auto`: This will return previews if the display options for that column is set to show the preview (see #944)
        - `all`: This will return previews for all columns
        - `none` (this is the default value): This will not return any previews
    - This would allow the front end to _always_ request the preview and toggle visibility of the preview without any additional requests. It also allows clients not to request the preview if it is not needed.
    - This would also open the door to other subtle UX enhancements such as displaying a record preview within a popover when the preview is disabled.
- The implementer should spec out the API response schema changes for this issue and get feedback from others on the team before implementing this issue.

## Additional context
- In #944 we are deciding how the record preview settings will be stored.
- [Foreign Key Constraints Spec](https://wiki.mathesar.org/en/design/specs/foreign-key-constraints) discusses the record preview in Scenario 4.

## Status
- Blocked by #944 because the back end will need to know which columns to return and whether to return them.

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #946 

Adds an additional field `previews` to `/records/` API which contains the preview data for the foreign key columns. 

Previews can be conditionally fetched by setting the `fk_previews` parameter to either `all` or `auto`. By setting it to `auto` previews will be fetched only for the columns whose display option `show_fk_preview` is set to `True` 

New API response for `/records/`

```

```json5
//Example schema
{
  "count": 1393,
  "grouping": {},
  "preview_data": [
    {
      "column": 100, // "Items"."Publication Id" colum ID
      
      // The Publications summary template: "Title: {278}"
      // The Authors summary template is: "{280} {281}"
     //
     // `100` is the foreign key column id in the items table, it is constrained to column id `277` which is the `id` column of the Publication Table. Column `278` is the title column in the publication table which is part of the preview template of the Publication Table.
     
      "template": "Title: {100__277__col__278}",
      "data": [{"100__277__col__278": "Blink"}]
      }
   ],
  "results": [
    {"10": 1, "20": 3.50, "100": 23},
  ]
  
}
```

## Additonal Context

The PR has few commented-out fixtures, due to the bug with the Data Explorer. I haven't removed it from the PR, as those give an idea of how the preview templating works as well as its edge cases.

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/records/operations/select.py [hunk 1]
```diff
@@ -67,6 +73,50 @@ def get_record(table, engine, id_value):
...
+    for record in records:
+        for referent_table_name, referent_obj in preview_columns.items():
+            constraint_columns = referent_obj['constraint_columns']
+            for constraint_column in constraint_columns:
+                constrained_column = constraint_column['constrained_column']
+                constrained_column_value = record[constrained_column]
+                if constrained_column_value is not None:
+                    referent_column = constraint_column['referent_column']
+                    preview_filters[referent_table_name][referent_column].add(constrained_column_value)
+    return preview_filters
```
Comment (mathemancer) [inline db/records/operations/select.py:117]: You should try using a join for this. It will be more readable and performant.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Please consider using joins to add the preview data in. It would be more idiomatic, and also brings performance benefits.

**Review by mathemancer — APPROVED** [review]
After a call with @silentninja , I agree we should get this merged in, then work out any performance issues or other changes later.

**Comment by mathemancer** [comment]
@dmos62 's PR merged right before my approving review. Thus, @silentninja needs to sort out some merge conflicts before we can get this merged to master.

**Comment by silentninja** [comment]
I am going to make the changes to the API just as we discussed. 

```json5
//Example schema
{
  "count": 1393,
  "grouping": {},
  "previews_render": [
    {
      "column": 10 // "Items"."Id",
      "template": "{10}"
    },
    {
      "column": 20 // "Items"."Acquisition Date",
      "template": "{20}"
    }
    {
      "column": 100, // "Items"."Publication Id" colum ID
      
      // The Publications summary template: "Title: {278}, Author: {279}"
      // The Authors summary template is: "{280} {281}"
      //
      // We use both those values to form the following template, putting the Authors
      // template inside the Publications template.
      "template": "Title: {278}, Author: {280&1} {281&1} Secondary Author: {280&2}",
      "data": [{"value" 23", "278": "Blink", "280&1": "Malcom", "281&1": "Gladwell", "280&2": "Sally"}]
  "results": [
    {"10": 1, "20": 3.50, "100": 23},
  ]
  
}
```

The field names might change, but the schema should remain the same

**Review by mathemancer — APPROVED** [review]
I think this looks good to merge. I'll try to submit a PR at some point that uncomments some tests once the data explorer bug is fixed.

**Pull request merged**